Add webhook support for work package comments and internal comments

wielinde opened the pull request:
# Ticket

https://community.openproject.org/wp/64129

# What are you trying to accomplish?
Allow webhook implementors to notice when a comment was added to a work package.

## Screenshots

![image](https://github.com/user-attachments/assets/84fe9a32-5fd3-43b7-bf5e-627d601e9cac)

# What approach did you choose and why?
This PR adds a new webhook resource called `WorkPackageComment` to which a webhook can register. It has the actions `comment` and `internal_comment`. The payload is equivalent to what would be rendered via the api v3 activity resource. It renders the JSON as via the ActivityRepresenter that would also be called on a simple API request.

# Merge checklist

- [x] Added/updated tests
- [x] Tested webhooks manually

wielinde committed:
Move triggering of work package comment web hooks to own resource

This decouples triggering comment webhooks from work package webhooks.
Thus a workpackage update webhook will still fire, even if the same
update resulted in a comment webhook.

NobodysNightmare commented:
@wielinde I restructured this to have a separate class handling the comments. In UI this means it's now listed separately:

![image](https://github.com/user-attachments/assets/84fe9a32-5fd3-43b7-bf5e-627d601e9cac)

(yes, I noticed the weird translations and will look into that soon)

In terms of behaviour this means that a work package update will always trigger an event, even if it includes a comment. I think that's less likely to break existing integrations. E.g. a webhook listening for changes to assignee should always notice these changes, regardless of whether I comment why I assigned a person or not.

wielinde commented:
@NobodysNightmare If I understand you correctly, when all four events are registered, a comment will now fire two requests, one for the comment and one for this actually being a work package update? Curios question: Is that cool or should we try to keep it at one, the more specific one? Is that typical behavior for webhook architectures? I am not an expert here. For me this solution definitely will solve the original requirements.

NobodysNightmare commented:
>  a comment will now fire two requests, one for the comment and one for this actually being a work package update?

yes, exactly

> Is that cool or should we try to keep it at one, the more specific one?

If one of them was simply the "superior webhook call", I'd be fine with just firing one of them. The problem I see is that none of them really is. Let's quickly think about consumers that only register for a work package update **or** only for the comments:

If you are subscribing for the updates, you want to track something like the status of WPs or any other attributes. The activity::comment payload would not be what you are looking for, because the payload is incompatible with what you expect from a WP update. Likewise if you look for comments, the WP update is lacking an important bit: The comment itself.

**In short:** Both events offer different information for different target audiences. Thus I think it's fine that we fire both of them.

NobodysNightmare commented:
I am going to add further tests today and then this PR should be ready.

NobodysNightmare committed:
Move webhook specs to correct location

They were just put into lib instead of the expectable
subfolder. This became apparent after I added a spec to
a more specific location.

Also fixed one or two Rubocop offenses along the way (not all
of them).

NobodysNightmare commented:
@Reviewer: Remaining Rubocop offenses are on existing specs that I moved into a new location and where I fixed some other offenses in the process already.

I silenced one Rubocop offense in new code that suggested to use `delegate` instead of a method definition. However, since the affected method definition is a method designed to be overridden in subclasses, I found it clearer to keep it defined as a `def`.

Feel free to point out if you disagree.
